### ISSUE TITLE
Extending libE fields when points are regiven

### ISSUE THREAD

**Issue by jmlarson1** [Libensemble/libensemble#593]
I am implementing an example gen/alloc interplay that re-gives points that have already been evaluated (for example, when the `sim_f` output is stochastic, and refinement at a past `sim_id` is desired). 

The `libE_fields` in the `main` branch were insufficient for this use case:
```
libE_fields = [('sim_id', int),        # Unique id of entry in H that was generated
               ('gen_worker', int),    # Worker that generated the entry
               ('gen_time', float),    # Time (since epoch) entry was entered into H
               ('given', bool),        # True if entry has been given for sim eval
               ('returned', bool),     # True if entry has been returned from sim eval
               ('given_time', float),  # Time (since epoch) that the entry was given
               ('sim_worker', int),    # Worker that did (or is doing) the sim eval
```
For example, once a point was generated, `given` to a sim, `returned` from a sim, and then `given_back` to a persistent gen, if it was generated again, the `gen_time` would store this time, but the `alloc_f` could see that it was already marked as `returned` and wouldn't know if that was for the first or second generation. We hoped that perhaps adding a `returned_time` would help the `alloc_f` determine that the point was re-generated after it was `returned`. 

After multiple discussions, we settled on the keys: 
```
libE_fields = [('sim_id', int),             # Unique id of entry in H that was generated
               ('gen_worker', int),         # Worker that (first) generated the entry
               ('sim_worker', int),         # Worker that did (or is doing) the sim eval
               ('first_gen_time', float),   # Time entry was first entered into H from a gen
               ('gen_time', float),         # Time entry was last requested by a gen
               ('given_time', float),       # Time entry was last given to be evaluated
               ('returned_time', float),    # Time entry was last returned from sim eval
               ('given_back_time', float),  # Time entry was given back to a persistent gen worker
               ('gen_count', int),          # Number of times an entry has been requested by a gen
               ('given_count', int),        # Number of times an entry has been given for sim eval
               ('returned_count', int),     # Number of times an entry has been returned from sim eval
               ('given_back_count', int),   # Number of times an entry has been given back to a persistent gen
```
This works for the single gen case, but will not be sufficient if we have multiple gens request the same point: in such a case, the requesting workers would not even be recorded. Nevertheless, I think we should ignore this case for the time being.

Thoughts on the above keys? We are bringing `given_back` into the `libE_fields` as a support for persistent generators.

**Comment by jlnav** [comment]
Has this been satisfied by #590 , or are several of these new fields pending for future implementation?

**Comment by jmlarson1** [comment]
Not fully. We included some non-breaking changes in #590. Changing the names of default libE fields can happen after the current (minor) release.

**Comment by jmlarson1** [comment]
This "regiving" feature is more difficult to implement than we initially thought. We are not going to consider implementing this at this time. For now, we recommend that if a user require a re-evaluation of a given simulation, it is easiest if the gen merely makes  new `sim_id`s with the same values.
